arai60/valid-parentheses/step3.py

```diff
@@ -0,0 +1,20 @@
+class Solution:
+    def isValid(self, s: str) -> bool:
+        stack = []
+        bracket_pairs = {
+            '(': ')',
+            '{': '}',
+            '[': ']',
+        }
+        for c in s:
+            if c in bracket_pairs:
+                stack.append(c)
+            else:
+                if not stack:
+                    return False
+
+                open_bracket = stack.pop()
+                if not bracket_pairs[open_bracket] == c:
+                    return False
```

Comment:
cだけ見るより、「stackのtopと合わせてペアができるか」に注目したほうが条件分岐が少なくなっていいのではと思いました。

（Pythonに詳しくなく、動くコードかは怪しいです...）

```suggestion
  
            top = stack[-1]
            is_match = False
            match (top, c):
                case ('(', ')'):
                    is_match = True
                case ('{', '}'):
                    is_match = True
                case ('[', ']'):
                    is_match = True
            if is_match:
                stack.pop()
            else:
                stack.append(c)

```

Reply:
コメントありがとうございます。
これは好みの問題ですが、stackに閉じカッコを入れる場合(上のコードで `c in [')', '}', ']']` かつ `is_match` がfalse)は、その時点で全体としてのカッコ列の整合性が崩れるので `return false` すると頭の中のスタックが解放される感じがします。
例えば `(}((((()))))` とあったときに、3文字目以降は見なくていいよね、という感じです。

Reply:
@kazukiii 
なるほどです。場合によっては早めにinvalid判定できるんですね

ありがとうございます〜